include/algorithms/public/PCA.hpp

```diff
@@ -35,47 +36,66 @@ class PCA
     MatrixXd         X = (input.rowwise() - mMean.transpose());
     BDCSVD<MatrixXd> svd(X.matrix(), ComputeThinV | ComputeThinU);
     mBases = svd.matrixV();
-    mValues = svd.singularValues();
+    mExplainedVariance = svd.singularValues().array().square() / (X.rows() - 1);
```

Comment:
I don't follow why we need to now store the explained variance rather than the singular values. Looks like it could equally well be calculated on demand by `getExplainedVariance`?

Reply:
I guess the idea is to move the calculation to the training instead of every time you process or, worse, procesFrame, so in all it made more sense to me

Reply:
I'm uneasy about making a breaking change on that basis, especially if we don't actually know if there's a performance problem there yet (presumably the square root which is still in `process` and `processFrame` could dominate the cost). If there is a performance problem caused by doing it on demand, I'd be more inclined to cache the result during training rather than invalidate any saved PCA models that folk have already made.

Reply:
Note that the square root here is done on an array of size k, where as the operations for getting the explained variance work on the full dimensionality. The idea of keeping explainedVariance follows PCA in sklearn. There both the singular values and explained variance are kept. Here, I was not sure I would want to keep the singular values unless we have a use for them. Clearly, it is cheaper to keep and use the explained variance.
I did not weigh breaking current code against the added complexity as this is still in beta, but I am not too fussy about it, we can calculate the explained variance on the fly (or on loading the JSON in addition to training) if the priority is to keep the JSON format.